### PR TITLE
Detect duplicate layout transitions in barrier

### PR DESCRIPTION
**Pull request by jzulauf-lunarg**
Addresses #78 

Validate a subset of internally inconsistent image memory barriers in a single barrier/event call. Duplicate image layout transistions for the same image and subresource range will generate errors.  Intersecting subresource ranges are not detected.  Duplicate transitions with oldLayout UNDEFINED are not invalid, and not flagged.

### CODE REVIEW

**Comment by jzulauf-lunarg** [comment]
@chrisforbes  -- this is a subset of the internally inconsistent barriers that could be caught, but I think a useful one.

File: layers/buffer_validation.cpp [hunk 1]
```diff
@@ -514,15 +514,60 @@ bool ValidateBarrierLayoutToImageUsage(layer_data *device_data, const VkImageMem
...
+                    if ((entry.barrier->newLayout != img_barrier->oldLayout) &&
+                        (img_barrier->oldLayout != VK_IMAGE_LAYOUT_UNDEFINED)) {
+                        const VkImageSubresourceRange &range = img_barrier->subresourceRange;
+                        skip = log_msg(
+                            core_validation::GetReportData(device_data), VK_DEBUG_REPORT_ERROR_BIT_EXT,
+                            VK_DEBUG_REPORT_OBJECT_TYPE_COMMAND_BUFFER_EXT, HandleToUint64(cb_state->commandBuffer),
+                            "VUID-VkImageMemoryBarrier-oldLayout-01197",
+                            "%s: pImageMemoryBarrier[%u] conflicts with earlier entry pImageMemoryBarrier[%u]. Image 0x%" PRIx64
+                            " subresourceRange: aspectMask=%u baseMipLevel=%u levelCount=%u, baseArrayLayer=%u, layerCount=%u, "
+                            "you cannot transition the layout from %s when current layout is %s.",
```
Comment (tobine) [inline layers/buffer_validation.cpp]: Does it make sense to update the last sentence to make it more clear that conflicting transitions is the core issue? Would this be more clear/correct:

Conflicting barrier is attempting to transition the layout from %s when earlier barrier already transitioned to %s.

**Comment by jzulauf-lunarg** [comment]
@tobine  -- check rewording.  Clearer, it think.

**Comment by chrisforbes** [comment]
Hang on, taking a look now.

File: layers/buffer_validation.cpp [hunk 2]
```diff
@@ -514,15 +514,60 @@ bool ValidateBarrierLayoutToImageUsage(layer_data *device_data, const VkImageMem
     return skip;
 }
 
+// Scoreboard for checking for duplicate and inconsistent barriers to images
+struct ImageBarrierScoreboardEntry {
+    uint32_t index;
+    const VkImageMemoryBarrier *barrier;
```
Comment (chrisforbes) [inline layers/buffer_validation.cpp]: I think you want to be super clear that this thing's lifetime is always bounded by the call to CmdPipelineBarrier etc. I immediately started trying to prove whether this ptr could dangle.

File: tests/layer_validation_tests.cpp [hunk 3]
```diff
@@ -12075,13 +12075,39 @@ TEST_F(VkLayerTest, InvalidBarriers) {
     vkCmdPipelineBarrier(m_commandBuffer->handle(), VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT,
                          VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT, VK_DEPENDENCY_BY_REGION_BIT, 0, nullptr, 0, nullptr, 1,
                          &img_barrier);
+
+    // TODO: this looks vestigal or incomplete...
     m_commandBuffer->BeginRenderPass(m_renderPassBeginInfo);
 
     img_barrier.oldLayout = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Suspect someone intended to call CmdPipelineBarrier here?

File: tests/layer_validation_tests.cpp [hunk 4]
```diff
@@ -12075,13 +12075,39 @@ TEST_F(VkLayerTest, InvalidBarriers) {
...
     // Can't send buffer memory barrier during a render pass
     vkCmdEndRenderPass(m_commandBuffer->handle());
 
+    // Duplicate barriers that change layout
+    img_barrier.image = image.handle();
+    img_barrier.oldLayout = VK_IMAGE_LAYOUT_UNDEFINED;
+    img_barrier.newLayout = VK_IMAGE_LAYOUT_GENERAL;
+    VkImageMemoryBarrier img_barriers[2] = {img_barrier, img_barrier};
+
+    // Tranisitons from UNDEFINED  are valid, even if duplicated
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: typo here

**Comment by karl-lunarg** [comment]
@jzulauf-lunarg  You may wish to rebase from master to pick up new CI configs and update the branch to re-run CI.

**Pull request merged**